src/include/ASTree.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include <Token.h>
+#include <vector>
+#include <ios>
+#include <sstream>
+
+struct Statement { };
+
+//////////////////////////////////////////
+/// Expressions
+
+struct Expression : public Statement { };
+
+class ExpressionGroup : public Expression {
+    std::unique_ptr<Expression> m_expression;
+public:
+    explicit ExpressionGroup(std::unique_ptr<Expression> expression)
+            : m_expression(std::move(expression)) { }
+};
+
+class BinaryExpression : public Expression {
+    TokenType m_op;
+    std::unique_ptr<ExpressionGroup> m_left;
+    std::unique_ptr<ExpressionGroup> m_right;
+public:
+    BinaryExpression(TokenType op, std::unique_ptr<ExpressionGroup> left,
+                     std::unique_ptr<ExpressionGroup> right)
+            : m_op(op)
+            , m_left(std::move(left))
+            , m_right(std::move(right)) { }
+};
+
+class UnaryExpression : public Expression {
+    char m_op;
+    std::unique_ptr<ExpressionGroup> m_operand;
+public:
+    UnaryExpression(char op, std::unique_ptr<ExpressionGroup> operand)
+            : m_op(op)
+            , m_operand(std::move(operand)) { }
+};
+
+class FunctionCall : public Expression {
+    std::string m_functionName;
+    std::vector<std::unique_ptr<ExpressionGroup>> m_args;
+public:
+    FunctionCall(std::string &functionName,
+                 std::vector<std::unique_ptr<ExpressionGroup>> args)
+            : m_functionName(functionName)
+            , m_args(std::move(args)) { }
+};
+
+//////////////////////////////////////////
+/// Literals
+
+class Literal : public Expression { };
+
+class Integer : public Literal {
+    int m_value;
+public:
+    explicit Integer(std::string value) : m_value(std::stoi(value)) { }
```

Comment:
How will we handle errors here?

Reply:
haven't thought too much about error handling, though we probably do want to at some point. I just assumed that anything the lexer says is an `Integer` would, in fact, be a valid integer:

```cpp

std::unique_ptr<Literal> Parser::parseLiteral() {
    std::unique_ptr<Literal> statement;
    switch (m_currentToken.type) {
        case TokenType::Integer:
            statement = std::make_unique<Integer>(
                    Integer(m_currentToken.literal));
            break;
        case TokenType::Float:
            statement = std::make_unique<Float>(
                    Float(m_currentToken.literal));
            break;
        case TokenType::String:
            statement = std::make_unique<String>(
                    String(m_currentToken.literal));
            break;
        case TokenType::True:
        case TokenType::False:
            statement = std::make_unique<Boolean>(
                    Boolean(m_currentToken.literal));
            break;
        default:
            // todo : exceptions, details
            printf("your token is shit");
            return std::unique_ptr<Literal>(nullptr);
    }
    getNextToken();
    return std::move(statement);
}

```